TL-3.1-HSBM.py

```diff
@@ -0,0 +1,291 @@
+        elif self.is_root is False:
+            if self.reembed is True:
+                n_components = self._choose_ncomponents()
+                X = self._embed(n_components=n_components)
+            else:
+                X = self.X
+
+        if np.linalg.norm(X) < 1:
+            X = X / np.linalg.norm(X)
+
+        # self.X_ = X
+        results = self._cluster(
+            X,
+            cluster_method="GMM",
+            n_init=self.n_init,
+        )
+
+        self.results_ = results
+
+    def _to_adjacency(self):
+        if self.adj is None:
+            pass
+        # TODO: convert data matrix e.g. distance or similarity matrix
+        # to adjacency matrix
+
+    def _choose_ncomponents(self):
+        max_dim = 0
+        for j, node in enumerate(self.get_lowest_level()):
+            dim = select_dimension(node.X, n_elbows=node.n_elbows)[0][-1]
+            if max_dim < dim:
+                max_dim = dim
+
+        return max_dim
+
+    def _embed(self, n_components=None):
+        if hasattr(self, "adj"):
+            adj = self.adj
+        else:
+            adj = self._to_adjacency()
+
+        embed_adj = pass_to_ranks(adj)
+        if self.plus_c:
+            embed_adj += 1 / adj.size
+
+        if n_components is None:
+            n_components = self.n_components
+
+        if self.embed_method == "ASE":
+            embedder = AdjacencySpectralEmbed(
+                n_components=n_components, n_elbows=self.n_elbows
+            )
+            embed = embedder.fit_transform(embed_adj)
+        elif self.embed_method == "LSE":
+            embedder = LaplacianSpectralEmbed(
+                n_components=n_components,
+                n_elbows=self.n_elbows,
+                regularizer=self.regularizer,
+            )
+            embed = embedder.fit_transform(embed_adj)
+
+        if self.normalize:
+            row_norms = np.linalg.norm(embed, axis=1)
+            embed /= row_norms[:, None]
+       
+        return embed
+
+    def _cluster(self, embed, cluster_method=None, n_init=None):
+        if self.cluster_method == "GMM":
+            cluster = AutoGMMCluster(
+                min_components=1, max_components=self.max_clusters
+            )
+            cluster.fit(embed)
+            model = cluster.model_
+            bic_ = model.bic(embed)
+            lik_ = model.score(embed)
+            k = cluster.n_components_
+            pred = cluster.fit_predict(embed)
+
+        res = {
+            "bic": -bic_,
+            "lik": lik_,
+            "k": k,
+            "model": model,
+            "pred": pred,
+        }
+        # TODO: if cluster_method == "Spherical-KMeans":
+        # TODO: if cluster_method == "KMeans":
+
+        return res
+
+    def _model_metrics(self):
+        results = self.results_
+        model = results["model"]
+        pred = results["pred"]
+        _n_components = results["k"]
+
+        return model, pred, _n_components
+
+    def select_model(self, k=None):
```

Comment:
1. this function was also very hacky on my part and 2. this should all happen in one of the other functions like cluster/fit. maybe we should talk more about how to design this part